vim/core/oni-core-interop/plugin/init.vim

```diff
@@ -12,6 +12,44 @@ function OniNotify(args)
 endfunction
 
 function OniNotifyBufferUpdate()
+    if !exists("b:last_change_tick")
+
+        let b:last_change_tick = -1
+        " Initial Send of whole buffer
+        let b:last_change_tick = b:changedtick
+        let buffer_lines = getline(1,"$")
+        let context = OniGetContext()
+        call OniNotify(["buffer_update", context, buffer_lines])
+    endif
+
+    if !exists("b:last_cursor_line")
+        let b:last_cursor_line = 1
+        let b:last_change_tick = b:changedtick
+        let buffer_lines = getline(1,"$")
+        let context = OniGetContext()
+        call OniNotify(["buffer_update", context, buffer_lines])
+    endif
+    if b:changedtick > b:last_change_tick
+        let b:last_change_tick = b:changedtick
+        if mode() == 'i'
+            if b:last_cursor_line < line(".")
+                let buffer_lines = getline(b:last_cursor_line, line("."))
+            elseif b:last_cursor_line > line(".")
+            let buffer_lines = getline(line("."),b:last_cursor_line)
+            elseif b:last_cursor_line == line(".")
+                let buffer_lines = [getline(".")]
+            endif
+
+            let context = OniGetContext()
+            call OniNotify(["incremental_buffer_update", context, buffer_lines, line(".")])
+        else
+            let buffer_lines = getline(1,"$")
+            let context = OniGetContext()
+            call OniNotify(["buffer_update", context, buffer_lines])
+        endif
+        let b:last_cursor_line = line(".")
+    endif
+endfunction
```

Comment:
If it's blocking load, the means Neovim hit an error while executing it on startup. It might be worth trying to call it once Neovim is loaded to see if you can capture the error. 

Regarding `last_cursor_line` and `last_change_tick` - if we remove those, then in insert mode, we'd just send over the single line? And in normal mode, we'd send over the whole buffer? It might be good to remove those, so I can see what it looks like simplified without those. That does seem like it would give a nice perf boost when typing in insert mode!